client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/OperationDispatcher.java

```diff
@@ -0,0 +1,868 @@
+
+      return operation.executeOperations(identifyOperationTarget(cacheName, connectionFailedServers),
+            this::executeOnSingleAddress);
+   }
+
+   public CompletionStage<Channel> executeAddListener(ClientListenerOperation operation) {
+      return executeAddListener(operation, getBalancer(operation.getCacheName()).nextServer(Set.of()));
+   }
+
+   public CompletionStage<Channel> executeAddListener(ClientListenerOperation operation, SocketAddress target) {
+      // Unfortunately, we have to do this preemptively for the case when a listener has include current state
+      // as that will not complete until all initial events have been received
+      clientListenerNotifier.addDispatcher(ClientEventDispatcher.create(operation,
+            target, () -> {/* TODO s*/}, operation.getRemoteCache()));
+      // Note this is done here, to guarantee this is lambda is ALWAYS invoked in the event loop after completion
+      // This allows a retry with a different address to be protected from concurrent events being raised
+      operation.whenComplete((channel, t) -> {
+         if (t != null) {
+            log.errorf("Error encountered trying to add listener %s", operation.listener);
+            clientListenerNotifier.removeClientListener(operation.listenerId);
+         } else {
+            SocketAddress unresolvedAddress = ChannelRecord.of(channel);
+            if (unresolvedAddress != target) {
+               // we have to reinstall the dispatcher with the correct address
+               clientListenerNotifier.addDispatcher(ClientEventDispatcher.create(operation,
+                     unresolvedAddress, () -> {/* TODO s*/}, operation.getRemoteCache()));
+            }
+            addListener(unresolvedAddress, operation.listenerId);
+            clientListenerNotifier.startClientListener(operation.listenerId);
+         }
+      });
+
+      return executeOnSingleAddress(operation, target);
+   }
+
+   private Function<Object, SocketAddress> identifyOperationTarget(String cacheName, Set<SocketAddress> failedServers) {
+      CacheInfo info = getCacheInfo(cacheName);
+      if (info != null && info.getConsistentHash() != null) {
+         ConsistentHash ch = info.getConsistentHash();
+         return ch::getServer;
+      }
+
+      FailoverRequestBalancingStrategy frbs = getBalancer(cacheName);
+      return obj -> frbs.nextServer(failedServers);
+   }
+
+   private <E> CompletionStage<E> execute(HotRodOperation<E> operation, Set<SocketAddress> opFailedServers) {
+      Object routingObj = operation.getRoutingObject();
+      SocketAddress targetAddress = null;
+      if (routingObj != null) {
+         CacheInfo cacheInfo = getCacheInfo(operation.getCacheName());
+         if (cacheInfo != null && cacheInfo.getConsistentHash() != null) {
+            SocketAddress server = cacheInfo.getConsistentHash().getServer(routingObj);
+            if (server != null && !opFailedServers.contains(server)) {
```

Comment:
Honestly, I am not sure. That is just how the code was before. I don't think it is possible though.. I think it might have been possible before scattered cache, but that is gone now.